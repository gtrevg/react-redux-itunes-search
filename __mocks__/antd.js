import React, { Component } from 'react';
import { mockComponent } from '../src/utils/testUtils';

export class Table extends Component {
	render() {
		return mockComponent('Antd Table', this.props);
	}
}

class Search extends Component {
	render() {
		return mockComponent('Antd Search', this.props);
	}
}

export const Input = { Search };
