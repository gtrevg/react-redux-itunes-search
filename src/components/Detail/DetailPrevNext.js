import React from 'react';
import { Tooltip, Button, Icon } from 'antd';

const lookUpIdIndex = (id, dataSource) => {
	return dataSource.findIndex(item => item.key === id);
};

const getPreviousItem = (idIndex, dataSource) => {
	if (idIndex === 0) return null;
	return dataSource[idIndex - 1];
};

const getNextItem = (idIndex, dataSource) => {
	if (idIndex === dataSource.lenght - 1) return null;
	return dataSource[idIndex + 1];
};

const DetailPrevNext = props => {
	const { id, dataSource } = props;
	const idIndex = lookUpIdIndex(id, dataSource);
	if (idIndex >= 0) {
		const previousItem = getPreviousItem(idIndex, dataSource);
		const nextItem = getNextItem(idIndex, dataSource);
		return (
			<Button.Group size="large">
				<Tooltip
					placement="bottomRight"
					title={`${previousItem.title} by ${previousItem.artist}`}
				>
					<Button
						type="primary"
						style={{ width: '120px', margin: '10px' }}
						href={`/detail?id=${previousItem.key}`}
					>
						<Icon type="left" />
						Next
					</Button>
				</Tooltip>
				<Tooltip
					placement="bottomLeft"
					title={`${nextItem.title} by ${nextItem.artist}`}
				>
					<Button
						type="primary"
						style={{ width: '120px', margin: '10px' }}
						href={`/detail?id=${nextItem.key}`}
					>
						Previous
						<Icon type="right" />
					</Button>
				</Tooltip>
			</Button.Group>
		);
	} else {
		return null;
	}
};

export default DetailPrevNext;
