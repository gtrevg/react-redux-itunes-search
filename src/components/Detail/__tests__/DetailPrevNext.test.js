import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import TestUtils from 'react-dom/test-utils';
import { MemoryRouter } from 'react-router-dom';
import { Button } from 'antd';

import DetailPrevNext from '../DetailPrevNext';
import store from '../../../config/mockStore';

const dataSource = [
	{
		album: 'Bad',
		artist: 'Michael Jackson',
		cover:
			'http://is2.mzstatic.com/image/thumb/Music/v4/05/c4/1c/05c41ca0-b614-a91c-e63e-2fd9a59d19e8/source/60x60bb.jpg',
		genre: 'Pop',
		key: 559334764,
		lenght: 257766,
		price: '1.29 USD',
		release: '1987-08-31T07:00:00Z',
		title: 'Smooth Criminal',
	},
	{
		album: 'Thriller',
		artist: 'Michael Jackson',
		cover:
			'http://is4.mzstatic.com/image/thumb/Music/v4/ec/2b/17/ec2b170d-2ecb-50af-02eb-39f0decc2847/source/60x60bb.jpg',
		genre: 'Pop',
		key: 269573364,
		lenght: 294188,
		price: '1.29 USD',
		release: '1982-11-30T08:00:00Z',
		title: 'Billie Jean',
	},
	{
		album: 'The Essential Michael Jackson',
		artist: 'Michael Jackson',
		cover:
			'http://is2.mzstatic.com/image/thumb/Music127/v4/8a/65/be/8a65bef2-f23d-e43d-9124-f5e4293513f7/source/60x60bb.jpg',
		genre: 'Pop',
		key: 159294478,
		lenght: 320905,
		price: '1.29 USD',
		release: '1987-08-31T07:00:00Z',
		title: 'Man In the Mirror',
	},
];

describe('DetailPrevNext component', () => {
	const id = 269573364;
	const changeTrack = jest.fn();
	const initialEntries = [
		{
			pathname: '/detail',
			search: `?id=${id}`,
			state: { dataSource },
		},
	];
	const component = (
		<MemoryRouter initialEntries={initialEntries}>
			<DetailPrevNext id={id} changeTrack={changeTrack} />
		</MemoryRouter>
	);

	it('Renders without crashing', () => {
		const div = document.createElement('div');
		ReactDOM.render(component, div);
		ReactDOM.unmountComponentAtNode(div);
	});

	it('Snapshot matchs', () => {
		const tree = renderer.create(component).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('Check the previous button', () => {
		const tree = TestUtils.renderIntoDocument(component);
		const buttons = TestUtils.scryRenderedComponentsWithType(tree, Button);
		const prevButtonOnClick = buttons[0].props.onClick;

		tree.history.push = jest.fn();
		prevButtonOnClick();

		expect(tree.history.push).toBeCalled();
		const expectedParameter = {
			pathname: '/detail',
			search: `?id=559334764`,
			state: {
				dataSource,
			},
		};
		expect(tree.history.push).toBeCalledWith(expectedParameter);
		expect(changeTrack).toBeCalledWith(dataSource[0]);
	});

	it('Check the next button', () => {
		const tree = TestUtils.renderIntoDocument(component);
		const buttons = TestUtils.scryRenderedComponentsWithType(tree, Button);
		const nextButtonOnClick = buttons[1].props.onClick;

		tree.history.push = jest.fn();
		nextButtonOnClick();

		expect(tree.history.push).toBeCalled();
		const expectedParameter = {
			pathname: '/detail',
			search: `?id=159294478`,
			state: {
				dataSource,
			},
		};
		expect(tree.history.push).toBeCalledWith(expectedParameter);
		expect(changeTrack).toBeCalledWith(dataSource[2]);
	});
});
