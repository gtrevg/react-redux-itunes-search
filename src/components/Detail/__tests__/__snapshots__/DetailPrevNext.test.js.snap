// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailPrevNext component Snapshot matchs 1`] = `
<mocked
  name="ButtonGroup"
  props={
    Object {
      "size": "large",
    }
  }
>
  <mocked
    name="Tooltip"
    props={
      Object {
        "placement": "bottomRight",
        "title": "Smooth Criminal by Michael Jackson",
      }
    }
  >
    <mocked
      name="Button"
      props={
        Object {
          "ghost": false,
          "onClick": [Function],
          "style": Object {
            "margin": "10px",
            "width": "120px",
          },
          "type": "primary",
        }
      }
    >
      <mocked
        name="Icon"
        props={
          Object {
            "type": "left",
          }
        }
      />
      Next
    </mocked>
  </mocked>
  <mocked
    name="Tooltip"
    props={
      Object {
        "placement": "bottomLeft",
        "title": "Man In the Mirror by Michael Jackson",
      }
    }
  >
    <mocked
      name="Button"
      props={
        Object {
          "ghost": false,
          "onClick": [Function],
          "style": Object {
            "margin": "10px",
            "width": "120px",
          },
          "type": "primary",
        }
      }
    >
      Previous
      <mocked
        name="Icon"
        props={
          Object {
            "type": "right",
          }
        }
      />
    </mocked>
  </mocked>
</mocked>
`;
